boa/src/builtins/array/mod.rs

```diff
@@ -98,8 +99,93 @@ impl Array {
     const LENGTH: usize = 1;
 
     fn constructor(this: &Value, args: &[Value], context: &mut Context) -> Result<Value> {
-        // Set Prototype
+        // Delegate to the appropriate constructor based on the number of arguments
+        match args.len() {
+            0 => Array::construct_array_empty(this, context),
+            1 => Array::construct_array_length(this, &args[0], context),
+            _ => Array::construct_array_values(this, args, context),
+        }
+    }
+
+    /// No argument constructor for `Array`.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array-constructor-array
+    fn construct_array_empty(this: &Value, context: &mut Context) -> Result<Value> {
+        let prototype = context.standard_objects().array_object().prototype();
+
+        Array::array_create(this, 0, Some(prototype), context)
+    }
+
+    /// By length constructor for `Array`.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array-len
+    fn construct_array_length(
+        this: &Value,
+        length: &Value,
+        context: &mut Context,
+    ) -> Result<Value> {
+        let prototype = context.standard_objects().array_object().prototype();
+        let array = Array::array_create(this, 0, Some(prototype), context)?;
+
+        if !length.is_number() {
+            array.set_field("0".to_string(), Value::from(length));
+            array.set_field("length".to_string(), Value::from(1));
```

Comment:
I know the spec says to convert to `String`s, but this will just do unnecessary allocations, `0` for line 137 and `"length"` for line 138 should work fine.